packages/circuit-ui/components/Modal/Modal.tsx

```diff
@@ -15,139 +16,311 @@
-import {
-  AccessibilityError,
-  isSufficientlyLabelled,
-} from '../../util/errors.js';
+import dialogPolyfill from '../../vendor/dialog-polyfill/index.js';
+import { applyMultipleRefs } from '../../util/refs.js';
 import { clsx } from '../../styles/clsx.js';
+import type { ClickEvent } from '../../types/events.js';
+import { isEscape } from '../../util/key-codes.js';
+import { useI18n } from '../../hooks/useI18n/useI18n.js';
+import { deprecate } from '../../util/logger.js';
 
 import classes from './Modal.module.css';
+import { createUseModal } from './createUseModal.js';
+import {
+  getFirstFocusableElement,
+  hasNativeDialogSupport,
+} from './ModalService.js';
+import { translations } from './translations/index.js';
+
+type DataAttribute = `data-${string}`;
+export interface ModalProps
+  extends Omit<HTMLAttributes<HTMLDialogElement>, 'children'> {
+  /**
+   * Whether the modal dialog is open or not.
+   */
+  open: boolean;
+  /**
+   * Callback when the modal dialog is closed.
+   */
+  onClose?: () => void;
+  /**
+   * a ReactNode or a function that returns the content of the modal dialog.
+   */
+  children?:
+    | ReactNode
+    | (({ onClose }: { onClose?: ModalProps['onClose'] }) => ReactNode);
+  /**
+   * Text label for the close button for screen readers.
+   * Important for accessibility.
+   */
+  closeButtonLabel?: string;
+  /**
+   * Use the `immersive` variant to focus the user's attention on the dialog content.
+   * default: 'contextual'
+   * */
+  variant?: 'contextual' | 'immersive';
+  /**
+   * Prevent users from closing the modal by clicking/tapping the overlay or
+   * pressing the escape key. Default `false`.
+   */
+  preventClose?: boolean;
+  /**
+   * Enables focusing a particular element in the dialog content and override default behavior
+   */
+  initialFocusRef?: RefObject<HTMLElement>;
 
-const TRANSITION_DURATION = 300;
-
-type PreventCloseProps =
-  | {
-      /**
-       * Text label for the close button for screen readers.
-       * Important for accessibility.
-       */
-      closeButtonLabel?: never;
-      /**
-       * Prevent users from closing the modal by clicking/tapping the overlay or
-       * pressing the escape key. Default `false`.
-       */
-      preventClose: boolean;
+  /**
+   * @deprecated this props was passed to react-modal and is no longer relevant.
+   * Use preventClose instead. Also see https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role#required_javascript_features
+   */
+  hideCloseButton?: boolean;
+  [key: DataAttribute]: string | undefined;
+}
+
+export const ANIMATION_DURATION = 300;
+
+export const Modal = forwardRef<HTMLDialogElement, ModalProps>((props, ref) => {
+  const {
+    open,
+    onClose,
+    closeButtonLabel,
+    variant = 'contextual',
+    children,
+    className,
+    preventClose,
+    initialFocusRef,
+    hideCloseButton,
+    ...rest
+  } = useI18n(props, translations);
```

Comment:
Any component with default translations needs to accept the `locale` prop so developers can customize with language to use.